### ISSUE TITLE
Would there be a version that runs on the Nintendo Switch as a Homebrew

### ISSUE THREAD

**Issue by Tokobotenkai** [Vita3K/Vita3K#2471]
Just asking if there's any intention to port this to the Nintendo Switch instead of running from Android.

**Comment by CoffeeBrewer64** [comment]
Currently, you cannot, but this is a great idea!

Problems:

- Currently, Switch controllers have problems where the controllers are swapped
- Some libraries won't work. Will SDL even work?
- CPU. Would either Dynarmic or Unicorn work on the Switch?
- Performance. Some parts of the Switch _may_ underperform for what Vita3K needs.
- Resolution. The Vita's resolution would have to be upscaled, so the FPS would decrease.
- Building. Building the Switch would have to be automated, but I'm sure there is some library for doing that somewhere on GitHub.

I actually really like this idea! Playing Vita games on a Switch would be cool. You can already do it on a Steam Deck, so on a Nintendo console? Even better.

You **_could_** try to get Linux on your Switch and run Vita3K through that, but Vita3K native to the Switch would be nicer.

**Comment by Tokobotenkai** [comment]
I was thinking since you can run Android on Switch (https://wiki.switchroot.org/en/Android/Setup-10), which from there you install Vita3K Apk.

I don't really understand the code, but it seems that version of Android is patched to utilise the Switch hardware?

Alternatively, the L4T version of Ubuntu on the Switch is patched to utilise the Switch hardware, but Vita3K is x86, so I guess we could try Box86/64 but I am not sure if the performance overhead is significant.

**Comment by nn9dev** [comment]
Intention for native HOS? I don't think that's on the to-do list at the moment, but anyone familiar with switch development would be greatly welcomed to to so, and I invite anyone with such skills to join us in the discord.

And to clear a few things up:
- someone has in fact had Vita3K running under Android switchroot on a switch
- someone has in fact had Vita3K running under Linux on Switch hardware (on a Jetson I think, but still uses the Tegra X1 chip)
- Dynarmic works on arm64 hosts! This is the dynamic recompiler that Vita3K uses. This would be different from [vita2hos by xerpi](https://github.com/xerpi/vita2hos) which instead uses the Switch's 32-bit execution mode (so, a translation layer), where the CPU instructions would be executed directly on the CPU instead of recompiled.
- As far as I'm aware, SDL2 does work on switch

Might be forgetting something, but...

TL;DR, absolutely possible, but not on the main roadmap afaik. Contributions by anyone knowledgeable is more than welcome :)

**Comment by Macdu** [comment]
Right now you can run Vita3K on the Switch by either:
- Installing Android on your Switch and using the provided APK
- Install Ubuntu (or any other distro) on the Switch then build Vita3K for the it (for the Linux aarch64 architecture), it works and some people have done it but it requires some knowledge to do. I'm planning on adding an automated aarch64 Linux build in the future.

**Comment by gladiator1014** [comment]
> I've simplified a bit for whoever wants to run it on l4t bionic (ffmpeg , libtompcrypt) into test https://github.com/mrcmunir/Vita3K
> 
> ```
> sudo add-apt-repository ppa:ubuntu-toolchain-r/test
> sudo apt-get update && sudo apt-get upgrade .
> wget https://apt.llvm.org/llvm.sh
> chmod +x llvm.sh
> sudo ./llvm.sh 14
> Install dependences :
> sudo apt install git cmake ninja-build pkg-config libgtk-3-dev xdg-desktop-portal libssl-dev libstdc+±10-dev
> Install update sdl2 from :
> bash <( curl https://raw.githubusercontent.com/cobalt2727/L4T-Megascript/master/helper.sh ) "scripts/sdl2_install_helper.sh"
> git clone https://github.com/mrcmunir/Vita3K
> cd Vita3K
> git submodule --init --recursive 
> cmake --preset linux-ninja-clang14
> cmake --build build/linux-ninja-clang14 --config Release
> ```
> 
I know this old but after running the last of these commands on a hacked Switch Lite running Ubuntu Jammy 22.04 I get a "1 error generated. ninja: build stopped: subcommand failed." 

Any ideas on how to fix that or what I should be looking for to identify the error?

**Comment by NumanTF2** [comment]
> Currently, you cannot, but this is a great idea!
> 
> Problems:
> 
>     * Currently, Switch controllers have problems where the controllers are swapped
> 
>     * Some libraries won't work. Will SDL even work?
> 
>     * CPU. Would either Dynarmic or Unicorn work on the Switch?
> 
>     * Performance. Some parts of the Switch _may_ underperform for what Vita3K needs.
> 
>     * Resolution. The Vita's resolution would have to be upscaled, so the FPS would decrease.
> 
>     * Building. Building the Switch would have to be automated, but I'm sure there is some library for doing that somewhere on GitHub.
> 
> 
> I actually really like this idea! Playing Vita games on a Switch would be cool. You can already do it on a Steam Deck, so on a Nintendo console? Even better.
> 
> You **_could_** try to get Linux on your Switch and run Vita3K through that, but Vita3K native to the Switch would be nicer.

https://github.com/hrydgard/ppsspp/pull/12323 ppsspp exists for the nintendo switch so you could try to study the source code for that and implement fixes in vita3k for switch. libnx is a library to compile nros (nintendo switch homebrew apps).

**Comment by mdandash** [comment]
This L4T fork: https://github.com/skydark719/Vita3K-L4T/releases/tag/release works @mspykerez but there is no memory mapping options to fix the graphical issues like invisible characters or images/graphics not appearing... I am able to get all of these things to work with the llvmpipe drivers but with unplayable framerates, on TegraX1 or OpenGL the graphics issues exist at 30fps+ and on Android OpenGL crashes the app, Vulkan drivers run with graphics issues, this is with games that are green/working on the compatibility list

**Comment by mrcmunir** [comment]
You needed Update Vulkan 1.2 drivers  for Memory mapping working in the nintendo switch with Double buffering options for fixes some bugs in Gravity rush for example
For Ubuntu 24.04 image based.

```
1-
sudo apt update && sudo apt dist-upgrade -y

2-
echo "deb [signed-by=/etc/apt/keyrings/theofficialgman-L4T.asc] https://theofficialgman.github.io/l4t-debs/ l4t noble-32-7" | sudo tee /etc/apt/sources.list.d/theofficialgman-L4T-32-7.list

3-
sudo rm -f /etc/apt/preferences.d/00-switch-bsp-restrictions

4-
sudo apt update && sudo apt install -y -o Dpkg::Options::="--force-confdef" nvidia-bsp-32-7

5-
reboot
```
or disable Memory mapping if some game doesn't happy with double buffering .

**Comment by mspykerez** [comment]
I'd prefer having it native on the Switch HOS instead of having to run linux on it.
